aten/src/ATen/native/NamedTensor.cpp

```diff
@@ -16,5 +16,141 @@ Tensor set_names(const Tensor& self, optional<DimnameList> names) {
+    if (tensor_names[dim] != aligned_names[idx]) {
+      continue;
+    }
+    // We've found a None name in `shorter` and `longer`. If their absolute positions
+    // from the right are not equal, then aligning the two names would require
+    // changing the absolute position from right of one of the None names,
+    // violating condition 2 of our [Alignment rules].
+    //
+    // For example:
+    // *, c, a, b
+    //       *, a
+    // [*, a] is a subsequence of [*, c, a, b], but in order to align them,
+    // we'd have to move the * to create [*, c: 1, a, b: 1]
+    if (tensor_names[dim].is_wildcard() &&
+        tensor_sizes.size() - dim != aligned_names.size() - idx) {
+      report_moving_unnamed_dim_error(
+          tensor_names, aligned_names, /*is_aligning_two_tensors=*/false);
+    }
+    expanded_sizes[idx] = tensor_sizes[dim];
+    --dim;
+  }
+  if (dim != -1) {
+    report_not_a_subsequence_error(
+        tensor_names, aligned_names, /*is_aligning_two_tensors=*/false);
+  }
+
+  return expanded_sizes;
+}
+
+// [Alignment rules]
+// Aligns `tensor` to names with the following rules:
+// 1) Check that tensor.names is a subsequence (not necessarily contiguous) of `names`.
+// 2) Aligning tensor.names to names must not change the absolute position from the
+//    right of any unnamed dimension.
+//
+// is_aligning_two_tensors tunes the error message to better match the following cases:
+// 1) tensor.align_to(names)  (is_aligning_two_tensors=false)
+// 2) torch.align_tensors([tensor, other])  (is_aligning_two_tensors=true)
+static Tensor align(const Tensor& tensor, DimnameList names, bool is_aligning_two_tensors) {
+  std::vector<int64_t> expanded_sizes;
+  if (tensor.names().has_value()) {
+    expanded_sizes = aligned_size(
+        tensor.sizes(), *tensor.names(), names, is_aligning_two_tensors);
+  } else {
+    auto tensor_sizes = tensor.sizes();
+    expanded_sizes = aligned_size(
+        tensor_sizes,
+        default_names(tensor_sizes.size()),
+        names,
+        is_aligning_two_tensors);
+  }
+  auto result = tensor.set_names(nullopt).view(expanded_sizes);
+  at::internal_set_names_inplace(result, names);
+  return result;
+}
+
+Tensor align_to(const Tensor& tensor, DimnameList names) {
+  TORCH_CHECK(
+      names.size() >= tensor.dim(),
+      "Cannot align tensor with dims ",
+      tensor.names() ? tensor.names().value() : default_names(tensor.dim()),
+      " to a shorter list of dims ", names, ".");
+  return align(tensor, names, /*aligning_two_tensors=*/false);
+}
+
+static std::vector<Tensor> align_tensors_to(TensorList tensors, DimnameList names) {
+  std::vector<Tensor> result;
+  result.reserve(tensors.size());
+  for (const auto& tensor : tensors) {
+    result.emplace_back(align(tensor, names, /*is_aligning_two_tensors=*/true));
+  }
+  return result;
+}
+
+std::vector<Tensor> align_tensors(TensorList tensors) {
```

Comment:
can `TensorList` be any number of tensors? If so should `..._two_tensors` be named something else?

Reply:
If the tensors are not alignable, the error message tells the user two of the tensors we can't align in the list. Because of that, I think it's fine that that variable is named `..._two_tensors`.